In [1]:
FILE_PATH = "/home/shivraj-pg/v3_poetrygen_new/Files/v3_gitapress_final_anustubh.csv"

In [19]:
import pandas as pd
from skrutable.meter_identification import MeterIdentifier
import re
from tqdm import tqdm

In [3]:
df = pd.read_csv(FILE_PATH)

In [4]:
df.head()

,sa,hi,skr_out,syllable_count,skr_meter,meter_cd,comments,prompt,split
0,न कामाननुरुद्ध्येत दुःखं कामेषु वै रतिः प्राप्...,इसलिये कामनाओं या भोगोंकी वृद्धिके लिये आग्रह ...,"anuṣṭubh (1,2: pathyā; 3,4: pathyā)",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train
1,यज्ञशिष्टाशिनः सन्तो मुच्यन्ते सर्वकिल्बिषैः भ...,कारण कि यज्ञसे शेष बचे हुए अन्नको खानेवाले श्र...,"anuṣṭubh (1,2: pathyā; 3,4: pathyā)",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train
2,स देवप्रस्थमासाद्य सेनाविन्दोः पुरं प्रति बलेन...,वहाँसे सेनाविन्दुकी राजधानी देवप्रश्वमें आकर च...,"anuṣṭubh (1,2: pathyā; 3,4: pathyā)",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train
3,अहोरात्रेक्षणो दिव्यो वेदाड़ः श्रुतिभूषण: आज्य...,"दिन और रात्रि उनके नेत्र हैं, वे दिव्यस्वरूप ह...","anuṣṭubh (1,2: pathyā; 3,4: pathyā)",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train
4,धर्मज्ञश्न कृतज्ञश्न दयावान्‌ प्रियभाषण: मान्य...,"ये धर्मज्ञ, कृतज्ञ, दयावान्‌ और प्रिय भाषण करन...","anuṣṭubh (1,2: pathyā; 3,4: asamīcīnā, [na ca ...",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train


In [5]:
df.columns

Index(['sa', 'hi', 'skr_out', 'syllable_count', 'skr_meter', 'meter_cd',
       'comments', 'prompt', 'split'],
      dtype='object')

In [6]:
MI = MeterIdentifier()

In [7]:
input_string = df['sa'][0]

print(input_string)

verse = MI.identify_meter(input_string, from_scheme='DEV', resplit_option='resplit_lite') 

print(verse.meter_label)

न कामाननुरुद्ध्येत दुःखं कामेषु वै रतिः प्राप्यार्थमुपयुञ्जीत धर्मं कामान् विसर्जयेत् ॥
anuṣṭubh (1,2: pathyā; 3,4: pathyā)


In [8]:
print(verse.summarize())

lggllggl    {m: 12}    [8: ysgl]
gggglglg    {m: 14}    [8: mrlg]
gglllggl    {m: 12}    [8: tsgl]
gggglglg    {m: 14}    [8: mrlg]

     na     kā     mā     na     nu     ru  ddhye     ta
      l      g      g      l      l      g      g      l
    duḥ   khaṃ     kā     me     ṣu    vai     ra    tiḥ
      g      g      g      g      l      g      l      g
    prā    pyā   rtha     mu     pa     yu    ñjī     ta
      g      g      l      l      l      g      g      l
    dha   rmaṃ     kā     mā    nvi     sa    rja    yet
      g      g      g      g      l      g      l      g

anuṣṭubh (1,2: pathyā; 3,4: pathyā)



In [9]:
summary = verse.summarize()

In [12]:
summary.split('\n')[0:4]

['lggllggl    {m: 12}    [8: ysgl]',
 'gggglglg    {m: 14}    [8: mrlg]',
 'gglllggl    {m: 12}    [8: tsgl]',
 'gggglglg    {m: 14}    [8: mrlg]']

In [16]:
x = ""
for i in summary.split('\n')[0:4]:
    x+=i[:8]
    
print(x)

lggllgglgggglglggglllgglgggglglg


In [17]:
print(len(x))

32


In [20]:
df['lg-pattern'] = ""

for i, r in tqdm(df.iterrows(), total=df.shape[0]):
    input_string = r['sa']
    verse = MI.identify_meter(
        input_string,
        from_scheme='DEV',
        resplit_option='resplit_lite'
    )

    summary = verse.summarize()

    x = ""
    for line in summary.split('\n')[0:4]:
        x += line[:8]

    if len(x) == 32:
        df.at[i, 'lg-pattern'] = x
    else:
        df.at[i, 'lg-pattern'] = "ERROR\n" + x

100%|██████████| 27208/27208 [00:17<00:00, 1594.14it/s]


In [21]:
df.head()

,sa,hi,skr_out,syllable_count,skr_meter,meter_cd,comments,prompt,split,lg-pattern
0,न कामाननुरुद्ध्येत दुःखं कामेषु वै रतिः प्राप्...,इसलिये कामनाओं या भोगोंकी वृद्धिके लिये आग्रह ...,"anuṣṭubh (1,2: pathyā; 3,4: pathyā)",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train,lggllgglgggglglggglllgglgggglglg
1,यज्ञशिष्टाशिनः सन्तो मुच्यन्ते सर्वकिल्बिषैः भ...,कारण कि यज्ञसे शेष बचे हुए अन्नको खानेवाले श्र...,"anuṣṭubh (1,2: pathyā; 3,4: pathyā)",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train,glgglggggggglglgglgglgggglgglglg
2,स देवप्रस्थमासाद्य सेनाविन्दोः पुरं प्रति बलेन...,वहाँसे सेनाविन्दुकी राजधानी देवप्रश्वमें आकर च...,"anuṣṭubh (1,2: pathyā; 3,4: pathyā)",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train,lggglgglgggglglllglllggllglllglg
3,अहोरात्रेक्षणो दिव्यो वेदाड़ः श्रुतिभूषण: आज्य...,"दिन और रात्रि उनके नेत्र हैं, वे दिव्यस्वरूप ह...","anuṣṭubh (1,2: pathyā; 3,4: pathyā)",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train,lggglggggggllgllglgglgggglgglglg
4,धर्मज्ञश्न कृतज्ञश्न दयावान्‌ प्रियभाषण: मान्य...,"ये धर्मज्ञ, कृतज्ञ, दयावान्‌ और प्रिय भाषण करन...","anuṣṭubh (1,2: pathyā; 3,4: asamīcīnā, [na ca ...",32,Anuṣṭubh,Anuṣṭubh,NaN,Objective:\n\nGenerate a sanskrit verse in Anu...,train,gggllggllggllgllgggllggggglglgll


In [22]:
df['lg-pattern'].str.contains('ERROR', na=False).sum()

np.int64(0)

In [23]:
df['lg-pattern'].nunique()

21453

In [24]:
df.shape[0]

27208

In [25]:
counts = df['lg-pattern'].value_counts().head(10)
percent = df['lg-pattern'].value_counts(normalize=True).head(10) * 100

result = pd.DataFrame({
    'count': counts,
    'percent': percent
})

result

,count,percent
lg-pattern,,
lggglgggglgglglggggllggglggglglg,11,0.040429
glgglggggggllglgglgglgggllgglglg,10,0.036754
glgglgglglgglglgglgglggggglllglg,9,0.033079
gggglggglggglglggggglggggggglglg,8,0.029403
glgglgglgggglglglggglgggglgllglg,8,0.029403
glgglgggglgllgllglgllgggglgglgll,8,0.029403
glgglgggglgglglggggglgggglgglglg,8,0.029403
gggglgggllgglglggggglgggglgglglg,8,0.029403
gggglgglglgglglglggglggglggglglg,8,0.029403


In [27]:
pattern_counts = df['lg-pattern'].value_counts()

result = (
    pattern_counts
    .value_counts()
    .sort_index()
    .rename_axis('n')
    .reset_index(name='number_of_patterns')
)

result['number_of_samples'] = (
    result['n'] * result['number_of_patterns']
)

result['percentage_of_samples'] = (
    result['number_of_samples'] / len(df) * 100
)

result

,n,number_of_patterns,number_of_samples,percentage_of_samples
0,1,17578,17578,64.605998
1,2,2760,5520,20.288151
2,3,643,1929,7.089827
3,4,282,1128,4.145839
4,5,132,660,2.425757
5,6,30,180,0.661570
6,7,17,119,0.437371
7,8,8,64,0.235225
8,9,1,9,0.033079
9,10,1,10,0.036754


The table shows the **frequency distribution of metrical patterns** in the dataset. Most patterns are rare: **64.6% of samples have a unique pattern**, while **20.3% belong to patterns occurring twice**. Overall, about **84.9% of samples belong to patterns occurring no more than twice**, whereas highly recurrent patterns are very uncommon. The most frequent individual pattern occurs only **11 times**.


In [ ]:
TODO: Pada level lg pattern analysis